File: ios/Packages/RideLinkPlatform/Sources/RideLinkPlatform/Sync/ResyncPorts.swift
import Foundation
import RideLinkCore

/// `ResyncCoordinator`'s exact call surface on `ControlSessionManager`'s `ResyncRelay`, mirroring
/// `SyncPlaybackChannel`/`SyncSessionPort`'s narrow-seam pattern exactly and for the same reason:
/// `ResyncCoordinator` depends on a concrete production class in `RideLinkPlatform` with no reason
/// to know a test double exists.
public protocol ResyncChannel: Sendable {
    func setSink(_ sink: (any ResyncSink)?) async
    @discardableResult func send(_ message: ResyncMessage) async -> Bool
}

/// `ResyncCoordinator`'s exact call surface on `ControlSessionManager`. Session **lifecycle** is
/// deliberately not here, for the reason `SyncSessionPort` already records: `ControlSessionManager
/// .onEvent` is a single mutable callback slot on this platform, and the app's `SessionCoordinator`
/// already owns it — so it forwards `.connected` explicitly to `ResyncCoordinator.onConnected`,
/// exactly as it already does for `SharedLibraryCoordinator` and `SyncPlaybackCoordinator`.
public protocol ResyncSessionPort: Sendable {
    var channel: any ResyncChannel { get }
    func currentAuthGeneration() async -> Int64
    /// ADR-025 §1: see `SyncSessionPort`'s equivalent contract — the liveness half, re-proved
    /// immediately before a leader's `STATE_SNAPSHOT` reply is sent.
    func liveAuthenticatedGeneration() -> Int64?
}

/// Zero-behaviour-change wrapper reaching the manager's `ResyncRelay`.
public struct ControlSessionResyncChannel: ResyncChannel {
    private let manager: ControlSessionManager

    public init(manager: ControlSessionManager) { self.manager = manager }

    public func setSink(_ sink: (any ResyncSink)?) async { await manager.resyncRelay().setSink(sink) }

    @discardableResult
    public func send(_ message: ResyncMessage) async -> Bool { await manager.resyncRelay().send(message) }
}

/// Zero-behaviour-change wrapper — the app composition root's production call site.
public struct ControlSessionResyncPort: ResyncSessionPort {
    private let manager: ControlSessionManager
    public let channel: any ResyncChannel

    public init(manager: ControlSessionManager) {
        self.manager = manager
        channel = ControlSessionResyncChannel(manager: manager)
    }

    public func currentAuthGeneration() async -> Int64 { await manager.currentAuthGeneration }

    public func liveAuthenticatedGeneration() -> Int64? { manager.liveAuthenticatedGeneration() }
}

File: android/app/src/main/kotlin/com/ridelink/app/resync/ResyncPorts.kt
package com.ridelink.app.resync

import com.ridelink.core.resync.ResyncMessage
import com.ridelink.network.control.ControlEvent
import com.ridelink.network.control.ControlSessionManager
import com.ridelink.network.resync.ResyncRelay
import com.ridelink.network.resync.ResyncSink
import kotlinx.coroutines.flow.SharedFlow

/**
 * ADR-023 Amendment A3's coordinator-level test seam, applied to [ResyncCoordinator] exactly as
 * `library.TransferPorts` applies it to [com.ridelink.app.library.SharedLibraryCoordinator]:
 * [ResyncCoordinator] depends on [ControlSessionManager], a concrete production class in `network`
 * with no reason to know a test double exists, so its **exact** call surface is declared here as a
 * narrow interface with a zero-behaviour-change adapter. [com.ridelink.app.sync.SyncPlaybackCoordinator]
 * and [com.ridelink.app.library.SharedLibraryCoordinator] need no equivalent wrapping: both already
 * live in `app`, so a test constructs real instances of them directly (as
 * `SyncPlaybackTwoPeerTest`/`SharedLibraryCoordinatorProviderAuthorizationTest` already do).
 */
interface ResyncChannelPort {
    var sink: ResyncSink?

    suspend fun send(message: ResyncMessage): Boolean
}

/** [ResyncCoordinator]'s exact call surface on [ControlSessionManager]. */
interface ResyncSessionPort {
    val resync: ResyncChannelPort
    val events: SharedFlow<ControlEvent>
    val currentAuthGeneration: Long

    /** ADR-025 §1: see [com.ridelink.app.library.TransferSessionPort.liveAuthenticatedGeneration] — same contract. */
    val liveAuthenticatedGeneration: Long?
}

/** Zero-behaviour-change wrapper around an already-constructed [ResyncRelay] (its constructor is `internal` to `network`). */
internal class ResyncRelayAdapter(
    private val delegate: ResyncRelay,
) : ResyncChannelPort {
    override var sink: ResyncSink?
        get() = delegate.sink
        set(value) {
            delegate.sink = value
        }

    override suspend fun send(message: ResyncMessage): Boolean = delegate.send(message)
}

/** Zero-behaviour-change wrapper — `AppContainer`'s production call site. */
internal class ResyncSessionManagerAdapter(
    private val delegate: ControlSessionManager,
) : ResyncSessionPort {
    override val resync: ResyncChannelPort = ResyncRelayAdapter(delegate.resync)
    override val events: SharedFlow<ControlEvent> get() = delegate.events
    override val currentAuthGeneration: Long get() = delegate.currentAuthGeneration
    override val liveAuthenticatedGeneration: Long? get() = delegate.liveAuthenticatedGeneration
}
